### PR TITLE
feat: add prune instances

### LINKED ISSUE #991: Add a prune-instances command
### What needs to get done

* Add a new command, `prune-instances` that removes all non-base instances for the active provider using the ProviderService. The following parameters are optional:  
  * `--all-providers` to prune all providers  
  * `--provider=<provider>` to prune for a specific provider  
    * If this isn’t specified, use the same business logic that lifecycle commands use to determine the provider.  
  * `--include-base` to prune base instances  
* Unit tests  
* A testcraft spread test that runs this command to prune LXD and Multipass instances

### Why it needs to get done

To allow users to prune instances.

Requires https://github.com/canonical/craft-providers/issues/639
Requires https://github.com/canonical/craft-providers/issues/893
Required for https://github.com/canonical/charmcraft/issues/1042

### PR DESCRIPTION
**Pull request by Aeonoi**
- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `make lint && make test`?
- [ ] Have you added an entry to the changelog (`docs/reference/changelog.rst`)?

closes #991

---

### CODE REVIEW

File: craft_application/services/provider.py [hunk 1]
```diff
@@ -597,3 +597,40 @@ def prepare_instance(instance: craft_providers.Executor) -> None:
...
+        if all_providers:
+            try:
+                providers.append(self._get_provider_by_name("lxd"))
+                providers.append(self._get_provider_by_name("LXD"))
+            except RuntimeError:
+                emit.debug("LXD provider not available, skipping.")
+
+            try:
+                providers.append(self._get_provider_by_name("multipass"))
+                providers.append(self._get_provider_by_name("Multipass"))
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply craft_application/services/provider.py]: Agreed, `_get_provider_by_name` runs `lower().strip()`, so you can just call it once.

File: craft_application/services/provider.py [hunk 2]
```diff
@@ -597,3 +597,40 @@ def prepare_instance(instance: craft_providers.Executor) -> None:
...
+            except RuntimeError:
+                emit.debug("Multipass provider not available, skipping.")
+
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply craft_application/services/provider.py]: The reasoning here is correct, if you decide to do `_add_provider_if_available`, I make it a class level function rather than a function within a function.

File: craft_application/services/provider.py [hunk 3]
```diff
@@ -597,3 +597,40 @@ def prepare_instance(instance: craft_providers.Executor) -> None:
...
+        else:
+            providers.append(self.get_provider(name=provider_name))
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply craft_application/services/provider.py]: The reasoning is correct, but you could probably make this more readable if you structured it along the lines of 

```python

if provider_name:
  ...
elif all_providers:
 ...
else:
 # default
 providers.append(self.get_provider())
```

File: craft_application/commands/prune_instances.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    def _fill_parser(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument(
+            "--all-providers",
+            action="store_true",
+            help="Prune instances from all providers",
+        )
+        parser.add_argument(
+            "--provider",
+            help="Prune for a specific provider",
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply craft_application/commands/prune_instances.py]: +1 on the mutual exclusion

File: craft_application/commands/prune_instances.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            help="Prune base instances (templates)",
+        )
+
+    def run(self, parsed_args: argparse.Namespace) -> None:
+        """Run the prune-instances command."""
+        self._services.provider.prune_instances(
+            all_providers=parsed_args.all_providers,
+            provider_name=parsed_args.provider,
+            prune_templates=parsed_args.include_templates,
+        )
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply craft_application/commands/prune_instances.py:59]: Agreed, do you mind adding a spread test? I think it's sufficient to do it with just LXD under `tests/spread/testcraft`.

File: craft_application/commands/prune_instances.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        """
+    )
+
+    def _fill_parser(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument(
+            "--all-providers",
+            action="store_true",
+            help="Prune instances from all providers",
+        )
+        parser.add_argument(
```
Comment (mr-cal) [inline craft_application/commands/prune_instances.py]: Can you add [choices](https://docs.python.org/3/library/argparse.html#choices)? That will improve the UX by giving quick errors up front if a user provides something besides `lxd` or `multipass`.

It will also make your life easier in the `prune_instances` command, because you'll know the user declared a valid provider name.

**Review by mr-cal — COMMENTED** [review]
Nice start, I left most of my feedback on top of the copilot review.

**Comment by lengau** [comment]
@copilot create a child PR that fixes the linting issues

File: tests/spread/testcraft/prune-instances/task.yaml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+execute: |
+  # Create a LXD instance.
+  testcraft pull --use-lxd
+
+  # Verify the instance exists. The name should start with 'testcraft-prune-instances-test-'.
+  lxc list --format csv -c n | grep "^testcraft-prune-instances-test-"
+
+  # Prune instances.
+  testcraft prune-instances --all
+
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply tests/spread/testcraft/prune-instances/task.yaml:15]: Agreed
```suggestion
  # Prune instances.
  testcraft prune-instances --all-providers
```

File: tests/spread/testcraft/prune-instances/task.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  # Verify the instance is gone.
+  if lxc list --format csv -c n | grep "^testcraft-prune-instances-test-"; then
+    echo "Instance was not pruned!"
+    exit 1
+  fi
+
+restore: |
+  testcraft prune-instances --all
```
Comment (mr-cal) [inline tests/spread/testcraft/prune-instances/task.yaml]: Here too:
```suggestion
  testcraft prune-instances --all-providers
```

File: craft_application/commands/prune_instances.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            choices=["lxd", "multipass"],
+        )
+
+    def run(self, parsed_args: argparse.Namespace) -> None:
+        """Run the prune-instances command."""
+        self._services.provider.prune_instances(
+            all_providers=parsed_args.all_providers,
+            provider_name=parsed_args.provider,
+            prune_templates=parsed_args.prune_templates,
+        )
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply craft_application/commands/prune_instances.py:59]: Agreed, I think `include_templates` accidentally got removed when making the other args mutually exclusive.

File: craft_application/services/provider.py [hunk 10]
```diff
@@ -597,3 +597,54 @@ def prepare_instance(instance: craft_providers.Executor) -> None:
...
+            current provider.
+        :param provider_name: Optional name of the provider to prune (if all_providers
+            is False).
+        :param prune_templates: Whether to also prune templates (if supported by the
+            provider).
+        """
+        providers: list[craft_providers.Provider] = []
+
+        if provider_name:
+            providers.append(self.get_provider(name=provider_name))
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply craft_application/services/provider.py]: Agreed. Nice to see copilot is making the same suggestions I was going to make.

File: tests/unit/commands/test_prune_instances.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+import argparse
+
+from craft_application.commands import PruneInstancesCommand
+
+
+def test_prune_instances_run(app_metadata, mock_services):
+    parsed_args = argparse.Namespace(
+        all_providers=True,
+        provider="lxd",
+        prune_templates=True,
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply tests/unit/commands/test_prune_instances.py:27]: Agreed.

File: tests/unit/commands/test_prune_instances.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    )
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply tests/unit/commands/test_prune_instances.py:28]: Agreed - these argparse values don't reflect what a user would be able to provide at the command line. 

IIRC, manually defining `parsed_args = argparse.Namespace(...)` in the test bypasses code like the mutually exclusive args. To test that code, you have to mock `argv` and assert the error with `capsys` ([example](https://github.com/canonical/craft-application/blob/4b0e00ae40b338ad9f837ac18bf063afebd086a4/tests/unit/test_application.py#L582-L600)).

I think you can do this in 2 tests.  One capsys-based test that ensures the args are mutually exclusive and another test that verifies the namespace args are successfully passed to `provider.prune_instances(...)`.  

You already have the second test written, but I recommend adding parametrizing it with `@pytest.mark.parameterize()` such that the test ensures what you set in ` parsed_args = argparse.Namespace(...)` actually get passed to `provider.prune_instances(...)`.

File: tests/unit/services/test_provider.py [hunk 13]
```diff
@@ -1174,3 +1174,54 @@ def test_configure_instance_with_pro_skipped(mocker, provider_service):
...
+        if name == "multipass":
+            return mock_multipass
+        raise RuntimeError(f"Unknown provider: {name}")
+
+    monkeypatch.setattr(provider_service, "_get_provider_by_name", mock_get_by_name)
+
+    provider_service.prune_instances(all_providers=True, prune_templates=False)
+
+    mock_lxd.prune.assert_called_once_with(prune_templates=False)
+    mock_multipass.prune.assert_called_once_with(prune_templates=False)
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply tests/unit/services/test_provider.py:1222]: I don't think we need to care about this - we're able to assume `commands.prune_instances` is only providing lowercase provider names to this function.

File: tests/unit/services/test_provider.py [hunk 14]
```diff
@@ -1174,3 +1174,54 @@ def test_configure_instance_with_pro_skipped(mocker, provider_service):
...
+    mock_lxd = mock.MagicMock()
+    mock_lxd.name = "lxd"
+
+    monkeypatch.setattr(provider_service, "get_provider", lambda name: mock_lxd)
+
+    provider_service.prune_instances(provider_name="lxd", prune_templates=True)
+
+    mock_lxd.prune.assert_called_once_with(prune_templates=True)
+
+
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply tests/unit/services/test_provider.py:1190]: I think you can ignore this, as this test needs to be updated once https://github.com/canonical/craft-application/pull/1058/changes#r3120374251 is accomdated.

File: tests/unit/services/test_provider.py
Comment (mr-cal) [inline tests/unit/services/test_provider.py:1]: Can you add a third unit test for when `provider_name = None`?  It should call `get_provider()`, which you can mock to return `lxd`.

File: craft_application/commands/prune_instances.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Aeonoi) [reply craft_application/commands/prune_instances.py:59]: Hmm, would the `_fill_parser()` method add the `include_templates` or `prune_templates` flag?

**Comment by Aeonoi** [comment]
How would I go about testing for the `prune` method since the current  craft-provider version doesn't have the `prune` method. So I'm not sure if the `prune-instances` method is actually done right when running it locally.

**Comment by mr-cal** [comment]
@Aeonoi, you can update the pyproject.toml and run `uv sync` to update the lockfile.

Feel free to push that change to this PR.  If the tests succeed in this PR, then we can make a new craft-providers release and you can update the pyproject.toml to the new release.


Here's an example: https://github.com/canonical/craft-application/pull/1047/changes/f155f77be69863643043ae4582353843e46fcce7

**Comment by Aeonoi** [comment]
> Here's an example: f155f77

This worked, thanks.

**Comment by mr-cal** [comment]
Great! By the way, there is now a 3.6.0 release of craft-providers.

https://canonical-craft-providers.readthedocs-hosted.com/en/latest/reference/changelog/
